lib/experimental/html/class-wp-html-character-reference-transcoder.php

```diff
@@ -0,0 +1,1149 @@
+						case 'attribute':
+							$buffer .= substr( $input, $next, $at - $next );
+							continue 3;
+
+						case 'data':
+							$buffer .= substr( $group, $sub_at, $sub_length );
+							continue 3;
+					}
+				}
+			}
+
+			/*
+			 * Whether by failing to find a group or failing to find a name,
+			 * we have failed to match a character reference name, so we can
+			 * continue processing as if this is plain text and leave the
+			 * invalid character reference name in place.
+			 */
+			$buffer .= substr( $input, $next, $at );
+		}
+
+		if ( 0 === $at ) {
+			return $input;
+		}
+
+		if ( $at < $end ) {
+			$buffer .= substr( $input, $at );
+		}
+
+		return $buffer;
+	}
+
+	/**
+	 * When numeric character references indicate certain code points
+	 * within the C1 control characters, these substitutions are used.
+	 *
+	 * @see https://html.spec.whatwg.org/#numeric-character-reference-end-state
+	 *
+	 * @var string[]
+	 */
+	static $c1_replacements = array(
+		0x80 => "€",
+		0x82 => "‚",
+		0x83 => "ƒ",
+		0x84 => "„",
+		0x85 => "…",
+		0x86 => "†",
+		0x87 => "‡",
+		0x88 => "ˆ",
+		0x89 => "‰",
+		0x8A => "Š",
+		0x8B => "‹",
+		0x8C => "Œ",
+		0x8E => "Ž",
+		0x91 => "‘",
+		0x92 => "’",
+		0x93 => "“",
+		0x94 => "”",
+		0x95 => "•",
+		0x96 => "–",
+		0x97 => "—",
+		0x98 => "˜",
+		0x99 => "™",
+		0x9A => "š",
+		0x9B => "›",
+		0x9C => "œ",
+		0x9E => "ž",
+		0x9F => "Ÿ",
+	);
+
+	/**
+	 * Autogenerated table of named character references.
+	 *
+	 * Character reference names are grouped by their first two characters,
+	 * then each group is string encoding the names and substitutions in
+	 * that group.
+	 *  - The first byte represents the length of the following name.
+	 *  - The bytes pertaining to the name follow the first byte.
+	 *  - The next byte represents the length of the following substitution for the previous name.
+	 *  - The bytes pertaining to the substitution follow this byte.
+	 *  - The end of the group string indicates that there are no more names in the group.
+	 *
+	 * The format for this lookup table is the result of an attempt to minimize memory
+	 * overhead while maintaining fast lookup. Since every named character reference is
+	 * at least three letters long (shortest is one of `&GT`, `&LT`, `&gt`, `&lt`), we
+	 * take the first two non-`&` letters as a group, leaving an average of 3.5 names
+	 * per group with a median of 2 names per group.
+	 *
+	 * The small size of each group should preserve memory locality of the group names
+	 * while maintaining quick testing of each candidate in the group since we know the
+	 * entire group and all its data will occupy consecutive bytes in memory.
+	 *
+	 * @see https://html.spec.whatwg.org/#named-character-references
+	 * @see https://html.spec.whatwg.org/entities.json
+	 *
+	 * @var string[] named character reference information
+	 */
+	public static $lookup_table = array(
+		"AE" => "\x04lig;\x02Æ\x03lig\x02Æ",               // &AElig; &AElig
```

Comment:
I'm confused by this `$lookup_table`. Why not store it as `"AE" => ["lig;" => "Æ", "lig" => "Æ"]`? Is this a memory optimization? Here's how `_zend_string` is stored:

```c
struct _zend_string {
	zend_refcounted_h gc;
	zend_ulong        h;                /* hash value */
	size_t            len;
	char              val[1];
};
```

`len` is just like size byte you use, and `val[1]` is used to store the actual contents (the length 1 is a "struct hack"). On top of that there's a few integer allocations for the hash (computed on the first use) and the gc reference. So `zend_string` would cost us a few integer allocations more and yield less complex code. 

What are your thoughts on that?

Reply:
this was an optimization; see d663e70f8205f06e2437ec5280ea2220edcc0018 for the previous data structure. I looked at another, which was storing these as static properties of an auto-generated class using a `Character_Reference` class with `$name` and `$substitution` properties, but each one had its own costs.

you're right that Zend strings don't require more storage, but a couple things led me away from the array as you wrote it:
 - array access involves a few steps of memory indirection and we don't know if the second or third name in a group will be close or distant in memory. the packed string format I'm using currently should keep most candidate names for a group _within a single cache line_ and all name/substitution values except potentially a few very long ones at that. I haven't measured the impact of this, but at least we know we're preserving locality during iteration of the candidates and will avoid all further memory lookups.
 - array key lookup requires an additional hashing step which here I don't think will be necessary. we're still doing it for groups, but I'm trying to balance the different tradeoffs and minimize the search space as quickly as possible. the packed string is mostly this same structure you're proposing except we know it comes from a contiguous sequence of bytes vs. a group of pointer dereferences.

because it's all been somewhat fast I've only been looking at memory overhead and as before, arrays have surprised me how much they add. I thought through another few approaches whereby the substitutions and names were all found in a single string, but those didn't offer much and were much messier in an editor.

additionally I looked at packing a tree in a single string which could potentially reduce the size of the table dramatically by avoiding most duplicated letters; it would let us scan one letter at a time (and also avoid new allocations) but I didn't code it up yet for no reason other than time.

there are 2231 named character references and we know we'd need at least a few bytes for each one, let alone the substitutions which total 6420 bytes. if we can get this down to a couple extra bytes of overhead for each one we're still looking at around a minimum of 10KB vs. 135KB, within one order of magnitude, and I don't think it's bad to add that much code on the server.

I'm not exactly sure what this approach would be like, but maybe I'll do an exploration at some point. this seemed pretty much good enough for now 🤷‍♂️

Reply:
I just played with packing trie into a string for fun: https://gist.github.com/adamziel/a628fc2d63a7733f9fed0dc9c3c34705

Reply:
@adamziel given that your trie representation still has some issues it's hard for me to test, _but_, I also find it maybe much more difficult to comprehend when reading. I do admit that my packed structure is itself a bit difficult to see, but at least the groups are easy to scan to and most are short enough where you can see all the names and replacements.

I predict that given we are doing one array lookup in my code based on the first two characters, that it's sufficiently optimized over the more abstract tree where we have approximately as many lookups as we have letters in the character name (less an optimization for when only one name remains given the prefix). the string scanning within a group also should be prompt given that the names within a group should be very local in memory and read in the same burst.

that being said, and knowing that in most cases we won't have any character references in an attribute, I guess we could relax this packing and go back to what's in the original commit (or a slight improvement) and see if we care about the performance.

ironically I do find the packed string in some ways more legible than the noisier array. concerning memory though there's a big difference between `'AE' => "\x04lig;\x02Æ\x03lig\x02Æ"` and this…

```php
	'AE' => array( 'lig;', 'Æ', 'lig', 'Æ' ),
	'AM' => array( 'P;', '&', 'P', '&' ),
	…
```

Whereas the packed string is a single memory lookup where all the values coincide, the array version includes string values which might be at very different locations in memory, so we still have in the case of `&lig` at least four separate memory lookups once we arrive at the group, since each value in the array is going to be a pointer to some other location with the actual string data.

---

As a side-note I find that any optimization to make the semi-colon optional is a bit fraught because it's not the case that all semi-colons are optional or required. so as far as the spec goes, the semicolon might as well be a normal letter.

---

For memory limiting I wonder how the trie implementation would look after fixing it. We have to make sure it finds `&lig;` before it finds `&lig` and it has to know at each prefix if there are branches, but also if there's a termination.

For example, the trie as written encodes `AElig;` and `AElig` in the same structure…

```
array(52) {
  ["A"]=>
  array(16) {
    ["E"]=>
    array(1) {
      ["l"]=>
      array(1) {
        ["i"]=>
        array(1) {
          ["g"]=>
          string(2) "Æ"
        }
      }
    }
  }
}
```

it looks like the problem here is that it finds the version without the semicolon and when the path is empty it overrides the existing `[";"]=>string(2) "Æ"` that was there already.

I was able to easily modify the script so that it stores the terminators at each branch point, but that made the packed table increase to 26 KB, and this trie traversal looks relatively slow compared to the string comparing. is 110 KB, if we can fit it in there, worth the trie?